toffy/normalize.py

```diff
@@ -0,0 +1,252 @@
+        raise ValueError('No bin files found in {}'.format(run_dir))
+
+    if file_prefix + '_combined.csv' in files:
+        warnings.warn('removing previously generated '
+                      'combined {} file in {}'.format(file_prefix, run_dir))
+        os.remove(os.path.join(run_dir, file_prefix + '_combined.csv'))
+        files = [file for file in files if 'combined' not in file]
+
+    if len(bins) != len(files):
+        raise ValueError('Mismatch between the number of bins and number '
+                         'of {} files'.format(file_prefix))
+
+    # collect all metrics files
+    metrics = []
+    for file in files:
+        metrics.append(pd.read_csv(os.path.join(run_dir, file)))
+
+    # check that all files are the same length
+    if len(metrics) > 1:
+        base_len = len(metrics[0])
+        for i in range(1, len(metrics)):
+            if len(metrics[i]) != base_len:
+                raise ValueError('Not all {} files are the same length: file {} does not match'
+                                 'file {}'.format(file_prefix, files[0], files[i]))
+
+    metrics = pd.concat(metrics)
+
+    metrics.to_csv(os.path.join(run_dir, file_prefix + '_combined.csv'), index=False)
+
+
+def combine_tuning_curve_metrics(dir_list):
+    """Combines metrics together into a single dataframe for fitting a turning curve
+
+    Args:
+        dir_list (list): list of directories to pull metrics from
+
+    Returns:
+        pd.DataFrame: dataframe containing aggregates metrics"""
+
+    # create list to hold all extracted data
+    all_dirs = []
+
+    # loop through directories, and if present, multiple fovs within directories
+    for dir in dir_list:
+
+        # generate aggregated table if it doesn't already exist
+        for prefix in ['pulse_heights', 'channel_counts']:
+            if not os.path.exists(os.path.join(dir, prefix + '_combined.csv')):
+                combine_run_metrics(dir, prefix)
+
+        # combine tables together
+        pulse_heights = pd.read_csv(os.path.join(dir, 'pulse_heights_combined.csv'))
+        channel_counts = pd.read_csv(os.path.join(dir, 'channel_counts_combined.csv'))
+        combined = pulse_heights.merge(channel_counts, 'outer', on=['fovs', 'masses'])
+
+        if len(combined) != len(pulse_heights):
+            raise ValueError("Pulse heights and channel counts must be generated for the same "
+                             "mass ranges and fovs. However, the data the following does do not "
+                             "exactly match: {}".format(dir))
+
+        # add directory label and add to list
+        combined['directory'] = dir
+        all_dirs.append(combined)
+
+    # combine data from each dir together
+    all_data = pd.concat(all_dirs)
+    all_data.reset_index()
+
+    # create normalized counts column
+    subset = all_data[['channel_counts', 'masses']]
+    all_data['norm_channel_counts'] = subset.groupby('masses').transform(lambda x: (x / x.max()))
```

Comment:
if we're normalizing by channel counts, an option to use the 99th percentile (instead of the max) to normalize could be useful.

Reply:
Normally I would agree, but in this case since the max over the entire field, rather than an individual cell, it's fairly robust already.